rasterio/_base.pyx

```diff
@@ -40,7 +40,65 @@ def gdal_version():
     return ver_b.decode('utf-8')
 
 
-cdef class DatasetReader(object):
+def get_dataset_driver(path):
+    """Return the name of the driver that would be used to open the
+    dataset at the given path."""
+    cdef void *hds = NULL
+    cdef void *drv = NULL
+    cdef const char *drv_name = NULL
+    cdef const char *fname = NULL
+
+    name_b = path.encode('utf-8')
+    fname = name_b
+    try:
+        with CPLErrors() as cple:
+            hds = _gdal.GDALOpen(fname, 0)
+            cple.check()
+    except CPLE_OpenFailed as exc:
+        raise RasterioIOError(str(exc))
+
+    drv = _gdal.GDALGetDatasetDriver(hds)
+    drv_name = _gdal.GDALGetDriverShortName(drv)
+    drv_name_b = drv_name
+    driver = drv_name_b.decode('utf-8')
+    _gdal.GDALClose(hds)
+
+    return driver
+
+
+def driver_can_create(name):
```

Comment:
Looking at these two functions,  I wonder if they can be condensed down further.
Untested code, functions need better names:

```
def driver_supports_mode(driver_name, mode):
    """Return True if the driver supports the mode"""
    cdef void *driver = NULL
    cdef const char *driver_name = NULL
    cdef char **metdata = NULL

    name_b = name.encode('utf-8')
    driver_name = name_b
    driver = _gdal.GDALGetDriverByName(driver_name)
    if driver == NULL:
        raise ValueError("No such driver: %s", name)

    try:
        metadata = _gdal.GDALGetMetadata(driver, NULL)
        return bool(_gdal.CSLFetchBoolean(metadata, mode, 0))

    finally:
        if metadata:
            _gdal.CSLDestroy(metadata)

def driver_can_create(driver_name):
    return driver_supports_mode(driver_name, 'DCAP_CREATE')

def driver_can_create_copy(driver_name):
    return driver_supports_mode(driver_name, 'DCAP_CREATECOPY')
```